### ISSUE TITLE
Lib not catching input on OS X 10.14 (18A391)

### ISSUE THREAD

**Issue by ArchoLen** [haschdl/pLaunchController#1]
As the old Processing forum is down i try to reach you this way:

The lib is somehow not working in OS X. Running your EasyCam example gives an error notice. The sketch runs fine despite it is not catching any input from the LaunchControl at all. Custom message in try() is not printed. Nor any other value change can be traced or messages get printed to console. I can modify the sketch to run with same LaunchControl and  "The MidiBus" lib. Thats why I assume it is with your lib.

I paste you the full console below:
>2018-10-14 11:59:40.027 java[1347:36590] pid(1347)/euid(501) is calling TIS/TSM in non-main thread environment, ERROR : This is NOT allowed. Please call TIS/TSM in main thread!!!
>2018-10-14 11:59:40.518 java[1347:36590] WARNING: NSWindow drag regions should only be invalidated on the Main Thread! This will throw an exception in the future. Called from (
	0   AppKit                              0x00007fff4ee342e3 -[NSWindow(NSWindow_Theme) _postWindowNeedsToResetDragMarginsUnlessPostingDisabled] + 386
	1   AppKit                              0x00007fff4ee3168c -[NSWindow _initContent:styleMask:backing:defer:contentView:] + 1488
	2   AppKit                              0x00007fff4ee310b6 -[NSWindow initWithContentRect:styleMask:backing:defer:] + 45
	3   libnativewindow_macosx.jnilib       0x000000012d3603fe Java_jogamp_nativewindow_macosx_OSXUtil_CreateNSWindow0 + 398
	4   ???                                 0x000000011b979a88 0x0 + 4757887624
)
>Connected to Launch Control MIDI Input.
>Connected to Launch Control MIDI Output.

I really hope you find the time and energy to fix this. 
P.S: I also own a LaunchControl XL and a LaunchPad S. If you want to tweak your lib to run on these (if OS X bug can be resolved) I'm glad to test it like you offered back in the old forum.

Thank you for your effort!
Len

**Comment by haschdl** [comment]
@ArchoLen thanks for sharing your feedback! I develop in Windows but I’m able to test it in a Mac - give me a couple of days and I will post an update here.

**Comment by ArchoLen** [comment]
Sure! I would love to see the lib working in OS X. Reach out if you want to try to support the other Novation devices. Love to help.

**Comment by ArchoLen** [comment]
@haschdl Quick update: The errors in the console are (most likely) caused by the P3D render. I had the exact same statements in an P2D Sketch I worked on yesterday. I further used your lib in a sketch w/o any declared rendering engine. Despite the fact that no input where registered the java errors did not show in the console.

**Comment by haschdl** [comment]
hello! I just tested the examples in Mac OS 10.13.6, with Processing 3.4. 
Like you said, using P3D shows some warnings, but at least the library seems to work in terms of MIDI drivers etc. There were a few glitches with timing of updating variables, and default values not being set - those were registered in issues #2 and #3, and fixed in the [latest release](https://github.com/haschdl/pLaunchController/releases/tag/latest) -  the new release should be available in the Processing IDE tomorrow, but you can also download from the repo if you want to try out before that. 
Note that after the update this is the easiest way to set a default value and to attach a knob to a variable in your sketch:
```Java
  controller.getKnob(KNOBS.KNOB_1_HIGH).range(0, 255).defaultValue(125).variable("hue");
  controller.getKnob(KNOBS.KNOB_2_HIGH).range(0, 255).defaultValue(125).variable("sat");
  controller.getKnob(KNOBS.KNOB_3_HIGH).range(0, 255).defaultValue(125).variable("bri");
  controller.getKnob(KNOBS.KNOB_4_HIGH).range(100, 500).defaultValue(200).variable("size");
```

Using the current release, you would need the longer version:
```Java
  controller.getKnob(KNOBS.KNOB_1_HIGH).range(0, 255).defaultValue(125).plugTo(this).variable("hue");
```
Both versions above are better placed in `setup()`. As I described in #3, there is a bug by which the default value does not change the variable until you do something with the controller, like updating a knob. Let me know if this makes sense, and if you were able to make progress.

**Comment by haschdl** [comment]
@ArchoLen  in addition to previous message, you can try to remove and reinstall the library in Processing. I'm closing the issue, as there are no pending errors - if you get another error, please submit another issue - like you notice, it seems best to avoid the p3d engine when doing the tests, just to make sure you don't get mixed warnings/errors.
The latest version should be 1.0a:
![image](https://user-images.githubusercontent.com/9550197/47113661-e6bf0200-d259-11e8-8f43-326ec83ec3b4.png)

**Comment by ArchoLen** [comment]
Sadly enough nothing changed for me. Neither in 1.0a or 1.1
No matter wich example I try (or if try to include lib in custom code) no input is detected. Again the custom error message is not displayed but the console give me:

> Connected to Launch Control MIDI Input.
> Connected to Launch Control MIDI Output.

When I connect my device. Respective:

> Launch Control not connected!

If not connected. 

I use the same device with "the MidiBus" lib, in Resolume 6 and VDMX 5.

**Comment by haschdl** [comment]
Issue reopened. I believe it might be related to templates... When you connect your device, which of the "User" or "Factory" button is lit? Try to set to Factory and then restart your sketch. (Meanwhile I'm working on a better solution, but Github is down and I wasn't able to upload the new version)

**Comment by haschdl** [comment]
Issue reopened. I believe it might be related to templates... When you connect your device, which of the "User" or "Factory" button is lit? Try to set to Factory and then restart your sketch. (Meanwhile I'm working on a better solution, but Github is down and I wasn't able to upload the new version)

**Comment by ArchoLen** [comment]
@haschdl Excuse the delay I got a couple of other things to take care of.

I tried "User" and "Factory" many times know. No difference. I checked the firmware wich is Version 124 and therefore on the latest release. I load the Launch Control Editor form Novation and did reset everything. But still the whole Lib seems to malfunction.

In your example you have the custom error:
>  try {
    controller = new LaunchController(this);
  }
  catch(Exception e) {
    println("Unfortunately we could not detect that Launch Control is connected to this computer :(");
  }

If I start the sketch with no device connect I only get this message in terminal:
> Launch Control not connected!

I'm convinced that the whole 
>controller = new LaunchController(this); 

does not been executed properly by Processing on OS X. And we can exclude my controller as a source of the failure for now. I was not able to do the promised try on windows. I will force it during the next few days.

**Comment by haschdl** [comment]
hello @ArchoLen! Is this something you might still be interested in? I have switched to Mac now and the library is working fine. I think there might be a bug somewhere I couldn't detect. Ping me here if you want to me take a look again.

**Comment by haschdl** [comment]
I identified two causes which might explaing this issue. Since I will be actively working on #8 and #10 below, I will close this one. If @ArchoLen or other Mac OS users later find this issue, please try with a newer version of the library, or comment below.

1. The Java MIDI implementation for Mac OS. Essentially Sysex messages were not working at all, making the library not work properly (issue #8 , solved in version 2.0.0-alpha)

3. Renaming the devices in the OS. Currently the library has hardcoded the name of the device. This is a bad approach as users can renamed their devices in the OS settings (for example when they have multiple of the same device connected to the computer). This is not solved yet, but captured in issue #10

**Comment by ArchoLen** [comment]
Nice to see this moving and having the XL supported for the future. Keep on the good work!
